### PR TITLE
Custom assertion messages for Snowhouse

### PR DESCRIPTION
**Pull request by hobovsky**
Attempt to extend Snowhouse assertions to make them able to accept custom assertion messages.

General idea: overload of `Assert::That` which accepts a wrapper of a Snowhouse `Expression` with additional message string. `DefaultFailureHandler` accepts such additional string and adds it to assertion messages, if provided.

Example:

```cpp
Assert::That(
    shortestDistance(1, 2, 3), 
    WithMessage(
        EqualsWithDelta(4.242640687, 1e-9), 
        "Invalid result for input: x=1, y=2, z=3"
    )
);
```

```text
<FAILED::>Invalid result for input: x=1, y=2, z=3<:LF:>Expected: equal to 4.24264 (+/- 1e-09)<:LF:>Actual: 5.24264<:LF:>
```

Pros:
- Backwards compatible, should not modify behavior of existing code in any way.
- Should work with any Snowhouse expression, not only `Equals`.

Cons:
- The fork modifies files of Snowhouse and does not realize its goal just by using extension points available in the framework.
- Syntax is rather clumsy, because main goal was to modify original Snowhouse as little as possible (see "Potential improvements" below).
- Does not provide three-argument variant of `AssertThat` macro. Can be used only as `Assert::That`

TODOs:
- Provide additional overloads of `Assert::That` which would work with `bool`, fluent expressions, `const char*`, etc.
- Document.

Potential improvements:
- Assertion syntax could be improved in one of following ways:
  - Provide a variant of `AssertThat` macro which would accept three arguments. Macro overloading, yay!
  - Provide some better way to create a message wrapper, for example `EqualsWithDelta(4.242640687, 1e-9).WithMessage("Invalid result for input: x=1, y=2, z=3")`. Would require modification of `Expression`.

Credits:

Inspired by user [MikChan](https://www.codewars.com/users/5de83a82bdfc590018f30f6a) and [his idea](https://www.codewars.com/kumite/607e94693d4a0d0006c63c32?sel=607ecf00eea8bf000755007d).

### CODE REVIEW

**Comment by kazk** [comment]
Converting to draft because we should maintain a separate branch (not target `master`), and also we're using `headers-only` branch. I pushed [`headers-only-codewars`](https://github.com/codewars/snowhouse/tree/headers-only-codewars), so we can branch from there.

---

Can you explain why we need a wrapper? Can we overload `Assert::That` to accept `(actual, expected, message)` similar to how you did to `DefaultFailureHandler`? https://github.com/codewars/snowhouse/pull/2

I don't think we need to use `AssertThat` macro?

**Comment by hobovsky** [comment]
> Can you explain why we need a wrapper? Can we overload `Assert::That` to accept `(actual, expected, message)` similar to how you did to `DefaultFailureHandler`?

Because such overload is ambiguous. `Assert::That` already accepts a string parameter in this place, with different meaning. Note that currently, prototype of `That` is exactly:

```cpp
void That(const ActualType& actual, const ExpressionType& expression, const char* file_name = "", int line_number = 0)`.
```

Adding `message` at the place you proposed risks (highly unprobable, I do admit) that all existing calls to `Assert::That(function_under_test("foo"), Equals("bar"), "some_file_name_used_for_whatever_reason.cpp")`  will change their meaning.

So basically:
- we can declare that we do not care about existing code and redefine the meaning of currently existing three-argument calls, or
- we introduce a set of functions parallel to `That` with another name and additional `message` parameter at third position, or
- we take some precautions that overloads of `That` with newly introduced `message` argument do not clash with existing code.

I choose the third approach: keep the name (i.e. `That`), and avoid clashes with potentially existing calls by making types of arguments different enough to avoid ambiguity. If you prefer any other approach, or have some better idea, let me know.

**Comment by kazk** [comment]
Oh, I missed that, thanks for the detailed explanation.

Yeah, let's avoid the first approach.

The second approach would be something like `Assert::WithMessageThat(actual, expression, message)`.

Another way for the third approach is to wrap only the message:
```cpp
Assert::That(
    shortestDistance(1, 2, 3), 
    EqualsWithDelta(4.242640687, 1e-9),
    Message("Invalid result for input: x=1, y=2, z=3")
);
// FailureMessage(), Note()
```

**Comment by hobovsky** [comment]
Do you prefer any of these, or I just prepare both in one fork and see how they work?

EDIT: Hey, I did not think of wrapping only the message, but it's a nice idea because the third argument could be of a template parameter type conforming to some string supplier contract. I think I will pick this approach for further investigation.

**Comment by kazk** [comment]
I prefer only wrapping the message, but you can try other approaches if you want.

The required change is minimal, and the API feels more natural to me.

`AssertionMessage` struct (this should be a safe enough name, but open to suggestions) can be put in `assert.h`.
```cpp
struct AssertionMessage {
  // Prevent implicit conversion
  explicit AssertionMessage(const std::string& message)
    : m_message { message }
  {
  }
  const std::string m_message;
};
// maybe make the field private and add a getter like `AssertionException` struct and match the codebase style
```

```cpp
Assert::That(
  shortestDistance(1, 2, 3),
  EqualsWithDelta(4.242640687, 1e-9),
  AssertionMessage("Invalid result for input: x=1, y=2, z=3")
);
```

**Comment by hobovsky** [comment]
Closing because superseded by #3
